src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -684,19 +685,25 @@ internal sealed override bool IsInterface
             }
         }
 
-        internal override ManagedKind ManagedKind
+        internal override ManagedKind GetManagedKind(ref HashSet<DiagnosticInfo> useSiteDiagnostics)
         {
-            get
+            var managedKind = _flags.ManagedKind;
+            if (managedKind == ManagedKind.Unknown)
             {
-                var managedKind = _flags.ManagedKind;
-                if (managedKind == ManagedKind.Unknown)
-                {
-                    var baseKind = base.ManagedKind;
-                    _flags.SetManagedKind(baseKind);
-                    return baseKind;
-                }
-                return managedKind;
+                HashSet<DiagnosticInfo>? managedKindUseSiteDiagnostics = null;
+                managedKind = base.GetManagedKind(ref managedKindUseSiteDiagnostics);
+                ImmutableInterlocked.InterlockedExchange(ref _managedKindUseSiteDiagnostics, managedKindUseSiteDiagnostics.ToImmutableArrayOrEmpty());
+                _flags.SetManagedKind(managedKind);
             }
+
+            Thread.MemoryBarrier();
```

Comment:
What protection do you believe this is giving you here?

Reply:
> Does this mean that we will get say obsolete warnings when we use a struct in a place where we checked the unmanaged constraint? 

The intent is to ensure that below we are dealing with the latest value of the _managedKindUseSiteDiagnostics field, since we are reading and updating two fields separately.

---
In reply to: [460267538](https://github.com/dotnet/roslyn/pull/46114#discussion_r460267538) [](ancestors = 460267538)

Reply:
Don't think the `MemoryBarrier` will provide that. The barrier controls read / write reordering from within a thread. It doesn't do anything to guarantee the synchronization of writes from other threads for this thread. 

Stated another way. Lets say Thread 1 is the one which enters the `if` (managedKind ==)` above and T2 skips the if and goes straight to this line. In T1 the order of writes is indeed `_managedKindUseSiteDiagnostics` and then `_flags`. That does not mean the order in which the writes become visible to T2 is the same. The order in which the writes become visible could be `_flags` then `_managedKindUseSiteDiagnostics`. The `MemoryBarrier` doesn't help here because all it does is control the order of read / writes within T2. It doesn't control whether or not those reads are consistent with the writse from T1. 

It's possible I'm remembering part of this wrong but pretty much whenever u have to depend on the ordering of writes between threads there is usually this type of gap.

Reply:
> The barrier controls read / write reordering from within a thread. 

That is what we are going after. Not any syncronization

---
In reply to: [460274273](https://github.com/dotnet/roslyn/pull/46114#discussion_r460274273) [](ancestors = 460274273)

Reply:
I think there is a race then. Even if we use the barrier to ensure the thread can't reorder it's reads of `_flags` and `_managedKindUseSiteDiagnostics` we can't guarantee the visibility of the writes from the other thread are both visible.

Reply:
> It's possible I'm remembering part of this wrong but pretty much whenever u have to depend on the ordering of writes between threads there is usually this type of gap.

I think ideally we need a volatile read here, but, I guess we don't have one, for an immutable array

>I think there is a race then. Even if we use the barrier to ensure the thread can't reorder it's reads of _flags and _managedKindUseSiteDiagnostics we can't guarantee the visibility of the writes from the other thread are both visible. 

I was suggesting an alternative to check both values in the if above, this would guarantee that we have non-default value here


---
In reply to: [460275688](https://github.com/dotnet/roslyn/pull/46114#discussion_r460275688) [](ancestors = 460275688)

Reply:
Agree I think either of those would work.